File: nodesms/controllers/studentController.js
const studentService = require('../services/studentService');

exports.createStudent = async (req, res) => {
  try {
    const studentDto = await studentService.createStudent(req.body);
    res.status(201).json(studentDto);
  } catch (error) {
    console.error('Create Student Error:', error);
    res.status(500).json({ message: error.message, errors: error.errors });
  }
};


exports.getStudentByRollNo = async (req, res) => {
  try {
    const student = await studentService.getStudentByRollNo(req.params.rollNo);
    res.json(student);
  } catch (error) {
    res.status(404).json({ message: error.message });
  }
};

exports.addMarksToStudent = async (req, res) => {
  try {
    const studentDto = await studentService.addMarksToStudent(req.params.rollNo, req.body);
    res.json(studentDto);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

exports.getStudentMarks = async (req, res) => {
  try {
    const marks = await studentService.getStudentMarks(req.params.rollNo);
    res.json(marks);
  } catch (error) {
    res.status(404).json({ message: error.message });
  }
};

exports.calculateCgpa = async (req, res) => {
  try {
    const cgpa = await studentService.calculateCgpa(req.params.rollNo);
    res.json(cgpa);
  } catch (error) {
    res.status(500).json({ message: error.message });
  }
};

File: student-management/src/main/java/com/example/student_management/controller/StudentController.java
package com.example.student_management.controller;


import com.example.student_management.dto.*;

import com.example.student_management.services.StudentService;

import org.springframework.http.ResponseEntity;

import org.springframework.web.bind.annotation.*;


import java.util.List;


@RestController
@RequestMapping("/students")
public class StudentController {

    private final StudentService studentService;

    public StudentController(StudentService studentService) {
        this.studentService = studentService;

    }

    @PostMapping("/create")
    public ResponseEntity<StudentDto> createStudent(@RequestBody CreateStudentDto createStudentDto) {
        return ResponseEntity.ok(studentService.createStudent(createStudentDto));

    }

    @GetMapping("/{rollNo}")
    public ResponseEntity<StudentDto> getStudentByRollNo(@PathVariable Long rollNo) {
        return ResponseEntity.ok(studentService.getStudentByRollNo(rollNo));

    }

    @PostMapping("/{rollNo}/marks")
    public ResponseEntity<StudentDto> addMarksToStudent(@PathVariable Long rollNo,
                                                        @RequestBody InputMarksDto inputMarksDto) {
        return ResponseEntity.ok(studentService.addMarksToStudent(rollNo, inputMarksDto));

    }

    @GetMapping("/{rollNo}/marks")
    public ResponseEntity<List<SemesterMarkDto>> getStudentMarks(@PathVariable Long rollNo) {
        return ResponseEntity.ok(studentService.getStudentMarks(rollNo));

    }

    @GetMapping("/{rollNo}/cgpa")
    public ResponseEntity<CgpaDto> calculateCgpa(@PathVariable Long rollNo) {
        return ResponseEntity.ok(studentService.calculateCgpa(rollNo));

    }
}
